pageserver/src/layered_repository.rs

```diff
@@ -2063,7 +2073,7 @@ impl LayeredTimeline {
             let (key, lsn, value) = x?;
 
             if let Some(prev_key) = prev_key {
-                if key != prev_key && writer.is_some() {
+                if writer.is_some() {
```

Comment:
This change doesn't look safe. For example, if you create a layer with key range 0..10 and LSN range 1000..2000, you must write *all* values in that rectangle. With this change, you might stop after writing half of the values for key 9.

Reply:
It is actually not correct - data corruption during ketteq backup restore in my case is actually caused by this change.
But frankly speaking I still do not understand why.
I do not understand your explanation. Assume that layer with key range 1..10 contains WAL records:
```
[{key: 1, lsn:1000}, {key:1, lsn:1001}, {key: 5, lsn: 1100}, {key:9, lsn: 1100}, {key: 9, lsn: 1200}, {key: 10, lsn: 1800}]
```
Before this change we can split this layer into two: 
```
1..10 => [{key: 1, lsn:1000}, {key:1, lsn:1001}, {key: 5, lsn: 1100}, {key:9, lsn: 1100}, {key: 9, lsn: 1200}]
10..inf => [{key: 10, lsn: 1800}]
```
with this change I expect to get the following layers:
```
1..10 => [{key: 1, lsn:1000}, {key:1, lsn:1001}, {key: 5, lsn: 1100}, {key:9, lsn: 1100}]
9..inf => [{key: 9, lsn: 1200}, {key: 10, lsn: 1800}]
```

I could not understand from `compact_level0` code why can we miss something `{key: 9, lsn: 1200}` in this case.
`for x in all_values_iter` should iterate though all values in any case.
Yes, we will get overlapped by key layers in this case.
But isn't it what we pretend to support?

Reply:
In any case, presence of overlapped layers (i.e. `1..10` and `9..inf`) is 
the only possible way to solve the problem with too large layer size for layers with a lot of occurrences of the same key, isn't it?
If so, then this change is correct and the problem is somewhere else in handling overlapped layers.
If you agree with it, I will continue investigation of overlapped layer traversal.

Reply:
Also it is a pity, that this my PR passed all tests, although (as we see) it doesn't work correctly.
It means that most of our tests doesn't produce large enough data to trigger compaction.
I wonder whether we should run tests with smaller `checkpoint_distance` (i.e. 1MB instead of 128).

Reply:
> Also it is a pity, that this my PR passed all tests, although (as we see) it doesn't work correctly.

Good chance to add such a test and increase our test coverage. ;-)

>  I wonder whether we should run tests with smaller checkpoint_distance (i.e. 1MB instead of 128).

Yeah, maybe we should. 1 MB is maybe a bit extreme, but maybe 16 MB or something. Although I'm not sure if you would get a lot of compactions with even that.

Reply:
> In any case, presence of overlapped layers (i.e. `1..10` and `9..inf`) is the only possible way to solve the problem with too large layer size for layers with a lot of occurrences of the same key, isn't it? If so, then this change is correct and the problem is somewhere else in handling overlapped layers. If you agree with it, I will continue investigation of overlapped layer traversal.

There is another way. Currently, the code first decides on the LSN range of the layers it's going to create, and then it divides the key range so that the resulting files are close to the target size. The problem here is that you have a single key with so many different values at different LSNs in that range, that the layer becomes too large no matter how narrow the key range for that layer is. In such a scenario, we should split by LSN. In a drawing, we are currently doing the thing on the left, but we should do the thing on the right:

```
LSN                            LSN
▲                              ▲
│ ┌────┬────┬─────┐            │ ┌────┬────┬─────┐
│ │    │    │     │            │ │    │    │ 1GB │
│ │    │    │     │            │ │    │    │     │
│ │    │    │     │            │ │    │    ├─────┤
│ │    │    │     │            │ │    │    │     │
│ │ 1GB│ 1GB│ 3GB │            │ │ 1GB│ 1GB│ 1GB │
│ │    │    │     │            │ │    │    ├─────┤
│ │    │    │     │            │ │    │    │ 1GB │
│ │    │    │     │            │ │    │    │     │
│ └────┴────┴─────┘            │ └────┴────┴─────┘
│                              │
└──────────────────────►Key    └──────────────────────►Key
```

Reply:
Thank you! Now I see the problem.
I tried to specify exact LSN range in DeltaLayer::finish.
But it doesn't help. I will try to investigate and debug it more, but my be you already know what's wring with this approach.

Reply:
Two passes makes sense. I don't see how to do it in one pass.

Reply:
With new version of the patch, ketteq backup is restored correctly.
Please check that new logic of splitting layer inside one key is correct.

Reply:
Right now I split layer on the same key only if its size is twice larger than checkpoint_distance. It means that total size of records with this key is >= checkpoint_distance. The probability that remaining number of records with this key which doesn't fit in this layer is very small seems to be not so large. So we can perform split without second pass. Yes - layer size in this case may be two times more than specified. Or be very small. But it seems to be still better than make two passes

Reply:
>  The probability that remaining number of records with this key which doesn't fit in this layer is very small seems to be not so large

Isn't that exactly what happened here?

Reply:
> Isn't that exactly what happened here?

Sorry, I do not undestand.
Before this patch we have huge layers files (> 1Gb), not small ones.
After this patch we have:
```
-rw-r----- 1 knizhnik knizhnik 536903680 июн 11 20:01 000000000000000000000000000000000000-000000067F000032D2000000000000000002__000000000169ACE0-000000000CF083A0
-rw-r----- 1 knizhnik knizhnik  95051776 июн 11 20:02 000000000000000000000000000000000000-000000067F000032D200027C580100000000__000000001F9844B8
-rw-r----- 1 knizhnik knizhnik 271400960 июн 11 20:02 000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__000000001A8A0261-000000001CC140B9
-rw-r----- 1 knizhnik knizhnik 271663104 июн 11 20:02 000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__000000001CC140B9-000000001DF7C041
-rw-r----- 1 knizhnik knizhnik 271720448 июн 11 20:02 000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__000000001DF7C041-000000001F7BD2A1
-rw-r----- 1 knizhnik knizhnik 271310848 июн 11 20:02 000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__000000001F7BD2A1-0000000021CB9591
-rw-r----- 1 knizhnik knizhnik 271179776 июн 11 20:02 000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__0000000021CB9591-0000000023982B61
-rw-r----- 1 knizhnik knizhnik 271286272 июн 11 20:02 000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__0000000023982B61-0000000024AA2B21
-rw-r----- 1 knizhnik knizhnik 271335424 июн 11 20:03 000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__0000000024AA2B21-0000000026260B11
-rw-r----- 1 knizhnik knizhnik 271196160 июн 11 20:03 000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__0000000026260B11-00000000287665C9
-rw-r----- 1 knizhnik knizhnik 271015936 июн 11 20:03 000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__00000000287665C9-000000002A4240F9
-rw-r----- 1 knizhnik knizhnik 271204352 июн 11 20:03 000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__000000002A4240F9-000000002B3C73F1
-rw-r----- 1 knizhnik knizhnik 271097856 июн 11 20:03 000000000000000000000000000000000000-FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF__000000002B3C73F1-000000002C4F5981
-rw-r----- 1 knizhnik knizhnik 536895488 июн 11 20:03 000000067F000032D2000000000000000000-000000067F000032D2000000000000000002__000000001A8A0261-000000001E8315D8
-rw-r----- 1 knizhnik knizhnik 536903680 июн 11 20:01 000000067F000032D2000000000000000001-000000067F000032D2000000000000000002__000000000169ACE0-0000000013686430
-rw-r----- 1 knizhnik knizhnik 536911872 июн 11 20:02 000000067F000032D2000000000000000001-000000067F000032D2000000000000000002__000000000169ACE0-0000000018B73FC0
-rw-r----- 1 knizhnik knizhnik 121151488 июн 11 20:02 000000067F000032D2000000000000000001-000000067F000032D2000000000000000002__000000000169ACE0-000000001A8A0261
-rw-r----- 1 knizhnik knizhnik 536920064 июн 11 20:03 000000067F000032D2000000000000000001-000000067F000032D2000000000000000002__000000001A8A0261-0000000023901430
-rw-r----- 1 knizhnik knizhnik 536903680 июн 11 20:03 000000067F000032D2000000000000000001-XXX__0000000023901430-000000002B3C73F1.yEj7p4E6.temp
-rw-r----- 1 knizhnik knizhnik 268492800 июн 11 20:02 000000067F000032D2000000700000000000-000000067F000032D200010E1E0000000254__000000000169ACE0-000000001A8A0261
-rw-r----- 1 knizhnik knizhnik 268468224 июн 11 20:02 000000067F000032D200010E1E0000000254-000000067F000032D20001B68B00000003E7__000000000169ACE0-000000001A8A0261
-rw-r----- 1 knizhnik knizhnik 213917696 июн 11 20:02 000000067F000032D20001B68B00000003E7-030000000000000000000000000000000002__000000000169ACE0-000000001A8A0261
```

so there are no small files and no files large than 265*2Mb.

Yes, there is non zero probability that we have to create delta layer with size smaller then few kb.
But probability of it is relatively small. So if it is critical to avoid the problem with too larger layer files (because them may cause problems with uploading to S3), then this PR  does it.
But the right solution IMHO will be to avoid serialization of directory.

Reply:
I wasn't worried about having occasional small files, I agree that's acceptable. I thought that the idea is that as soon as soon as you notice that the current file is getting too large, you finish writing the records for the current key, and then start a new layer file. But that's how it already works in 'main', and that's not good enough because there might be so many records for the current key that you end up with an over-sized layer file.

I don't think this is possible with a single pass.